File: app/src/main/java/com/gnoemes/shikimoriapp/domain/anime/series/AlternativeSeriesInteractorImpl.java
package com.gnoemes.shikimoriapp.domain.anime.series;

import com.gnoemes.shikimoriapp.data.repository.alternative.AlternativeRepository;
import com.gnoemes.shikimoriapp.data.repository.app.UserSettingsRepository;
import com.gnoemes.shikimoriapp.data.repository.rates.UserRatesRepository;
import com.gnoemes.shikimoriapp.entity.anime.series.domain.AlternativeTranslation;
import com.gnoemes.shikimoriapp.entity.anime.series.domain.AlternativeTranslationType;
import com.gnoemes.shikimoriapp.entity.anime.series.domain.Episode;
import com.gnoemes.shikimoriapp.entity.app.domain.Type;
import com.gnoemes.shikimoriapp.entity.rates.domain.UserRate;
import com.gnoemes.shikimoriapp.utils.rx.CompletableErrorHandler;
import com.gnoemes.shikimoriapp.utils.rx.RxUtils;
import com.gnoemes.shikimoriapp.utils.rx.SingleErrorHandler;

import java.util.List;

import javax.inject.Inject;

import io.reactivex.Completable;
import io.reactivex.Single;

public class AlternativeSeriesInteractorImpl implements AlternativeSeriesInteractor {

    private AlternativeRepository repository;
    private UserRatesRepository ratesRepository;
    private UserSettingsRepository settingsRepository;
    private SingleErrorHandler singleErrorHandler;
    private CompletableErrorHandler completableErrorHandler;
    private RxUtils rxUtils;

    @Inject
    public AlternativeSeriesInteractorImpl(AlternativeRepository repository,
                                           UserRatesRepository ratesRepository,
                                           UserSettingsRepository settingsRepository,
                                           SingleErrorHandler singleErrorHandler,
                                           CompletableErrorHandler completableErrorHandler,
                                           RxUtils rxUtils) {
        this.repository = repository;
        this.ratesRepository = ratesRepository;
        this.settingsRepository = settingsRepository;
        this.singleErrorHandler = singleErrorHandler;
        this.completableErrorHandler = completableErrorHandler;
        this.rxUtils = rxUtils;
    }

    @Override
    public Single<List<Episode>> getEpisodes(long animeId) {
        return repository.getAnimeEpisodes(animeId)
                .compose((SingleErrorHandler<List<Episode>>) singleErrorHandler)
                .compose(rxUtils.applySingleSchedulers());
    }

    @Override
    public Completable setEpisodeWatched(long animeId, long episodeId, long rateId) {
        return repository.isEpisodeWatched(animeId, episodeId)
                .flatMapCompletable(isWatched -> {
                    if (!isWatched) {
                        return repository.setEpisodeWatched(animeId, episodeId)
                                .andThen(ratesRepository.onEpisodeWatched(rateId));
                    } else {
                        return Completable.complete();
                    }
                })
                .compose(completableErrorHandler)
                .compose(rxUtils.applyCompleteSchedulers());
    }

    @Override
    public Completable setEpisodeWatched(long animeId, long episodeId) {
        return repository.isEpisodeWatched(animeId, episodeId)
                .flatMapCompletable(isWatched -> {
                    if (isWatched) {
                        return repository.setEpisodeWatched(animeId, episodeId)
                                .toSingleDefault(settingsRepository.getUser())
                                .flatMapCompletable(user -> ratesRepository.createRate(animeId, Type.ANIME, UserRate.getStartWatching(), user.getId()));
                    } else {
                        return Completable.complete();
                    }
                })
                .compose(completableErrorHandler)
                .compose(rxUtils.applyCompleteSchedulers());
    }

    @Override
    public Single<List<AlternativeTranslation>> getEpisodeTranslations(AlternativeTranslationType type, long animeId, long episodeId) {
        return repository.getTranslations(type, animeId, episodeId)
                .compose((SingleErrorHandler<List<AlternativeTranslation>>) singleErrorHandler)
                .compose(rxUtils.applySingleSchedulers());
    }

    @Override
    public Completable clearHistory(long animeId) {
        return repository.clearHistory(animeId)
                .compose(completableErrorHandler)
                .compose(rxUtils.applyCompleteSchedulers());
    }
}

File: app/src/main/java/com/gnoemes/shikimoriapp/domain/anime/series/AlternativeSeriesInteractorImpl.kt
package com.gnoemes.shikimoriapp.domain.anime.series

import com.gnoemes.shikimoriapp.data.repository.alternative.AlternativeRepository
import com.gnoemes.shikimoriapp.data.repository.app.UserSettingsRepository
import com.gnoemes.shikimoriapp.data.repository.rates.UserRatesRepository
import com.gnoemes.shikimoriapp.entity.anime.series.domain.AlternativeTranslation
import com.gnoemes.shikimoriapp.entity.anime.series.domain.AlternativeTranslationType
import com.gnoemes.shikimoriapp.entity.anime.series.domain.Episode
import com.gnoemes.shikimoriapp.entity.app.domain.Type
import com.gnoemes.shikimoriapp.entity.app.domain.UserSettings
import com.gnoemes.shikimoriapp.entity.main.domain.Constants
import com.gnoemes.shikimoriapp.entity.rates.domain.UserRate
import com.gnoemes.shikimoriapp.utils.rx.CompletableErrorHandler
import com.gnoemes.shikimoriapp.utils.rx.RxUtils
import com.gnoemes.shikimoriapp.utils.rx.SingleErrorHandler
import io.reactivex.Completable
import io.reactivex.Single
import javax.inject.Inject

class AlternativeSeriesInteractorImpl @Inject constructor(
        private val repository: AlternativeRepository,
        private val ratesRepository: UserRatesRepository,
        private val settingsRepository: UserSettingsRepository,
        private val singleErrorHandler: SingleErrorHandler<Any>,
        private val completableErrorHandler: CompletableErrorHandler,
        private val rxUtils: RxUtils
) : AlternativeSeriesInteractor {


    override fun getEpisodes(animeId: Long): Single<List<Episode>> {
        return repository.getAnimeEpisodes(animeId)
                .compose(singleErrorHandler as SingleErrorHandler<List<Episode>>)
                .compose(rxUtils.applySingleSchedulers())
    }

    override fun setEpisodeWatched(animeId: Long, episodeId: Long, rateId: Long): Completable {
        return repository.isEpisodeWatched(animeId, episodeId)
                .flatMapCompletable { isWatched ->
                    if (!isWatched) {
                        repository.setEpisodeWatched(animeId, episodeId)
                                .andThen(createRateIfNeed(animeId))
                                .andThen(if (rateId != Constants.NO_ID) ratesRepository.onEpisodeWatched(rateId) else Completable.complete())
                    } else {
                        Completable.complete()
                    }
                }
    }

    override fun getEpisodeTranslations(type: AlternativeTranslationType, animeId: Long, episodeId: Long): Single<List<AlternativeTranslation>> =
            repository.getTranslations(type, animeId, episodeId)
                    .compose(singleErrorHandler as SingleErrorHandler<List<AlternativeTranslation>>)
                    .compose(rxUtils.applySingleSchedulers())

    override fun clearHistory(animeId: Long): Completable =
            repository.clearHistory(animeId)
                    .compose(completableErrorHandler)
                    .compose(rxUtils.applyCompleteSchedulers<Any>())

    private fun createRateIfNeed(animeId: Long): Completable {
        return settingsRepository.userSettings
                .filter(UserSettings::getAutoStatus)
                .filter { it.userBrief != null }
                .flatMapCompletable { ratesRepository.createRate(animeId, Type.ANIME, UserRate.getStartWatching(), it.userBrief.id) }
    }
}